cdap-app-templates/cdap-etl/cdap-data-streams-base/src/main/java/io/cdap/cdap/datastreams/SparkStreamingPipelineRunner.java

```diff
@@ -185,4 +214,124 @@ protected SparkCollection<Object> handleJoin(Map<String, SparkCollection<Object>
     shufflers.add(stageName);
     return handleJoin(joiner, inputDataCollections, stageSpec, functionCacheFactory, numPartitions, collector);
   }
+
+  @Override
+  protected void processDag(PhaseSpec phaseSpec, String sourcePluginType, JavaSparkExecutionContext sec,
+                            Map<String, Integer> stagePartitions, PluginContext pluginContext,
+                            Map<String, StageStatisticsCollector> collectors, PipelinePhase pipelinePhase,
+                            FunctionCache.Factory functionCacheFactory,
+                            MacroEvaluator macroEvaluator, CombinerDag groupedDag,
+                            Map<String, Set<String>> groups, Set<String> branchers,
+                            Set<String> shufflers) throws Exception {
+    if (nativeStateStoreEnabled) {
+      //Process source separately
+      List<String> topologicalOrder = groupedDag.getTopologicalOrder();
+      String source = topologicalOrder.get(0);
+      StageSpec stageSpec = pipelinePhase.getStage(source);
+      StageStatisticsCollector collector = collectors.get(source) == null ? new NoopStageStatisticsCollector()
+        : collectors.get(source);
+      JavaDStream<Object> dStream = getDStream(stageSpec, collector);
+      DataTracer dataTracer = sec.getDataTracer(stageSpec.getName());
+      dStream.foreachRDD((javaRDD, time) -> {
+        Transactionals.execute(sec, context -> {
```

Comment:
cdap transactions timeout after 30 seconds by default, so we would want to get the sinkRunnables inside the transaction, but actually execute them outside the transaction.

Reply:
Done